File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count)

(def nucleotides {\A 0 \T 0 \C 0 \G 0})

(defn nucleotide-counts [s]
  (->> s
       (frequencies)
       (#(conj nucleotides %))))

(defn count-of-nucleotide-in-strand [nt s]
  (if (nucleotides nt)
    ((nucleotide-counts s) nt)
    (throw (Exception. "Bad nucleotide"))))

File: elixir/nucleotide-count/lib/nucleotide_count.ex
defmodule NucleotideCount do
  @nucleotides [?A, ?C, ?G, ?T]

  @spec count(charlist(), char()) :: non_neg_integer()
  def count(strand, nucleotide) do
    strand |> Enum.count(&(nucleotide == &1))
  end

  @spec histogram(charlist()) :: map()
  def histogram(strand) do
    @nucleotides |> Enum.into(%{}, fn n -> {n, count(strand, n)} end)
  end
end
